bsdbpf/bsd_bpf_sniffer.go

```diff
@@ -0,0 +1,199 @@
+	BPFDeviceName string
+	// ReadBufLen specifies the size of the buffer used to read packets
+	// off the wire such that multiple packets are buffered with each read syscall.
+	// The packet will be clipped if it's size exceeds the buffer size.
+	// A larger buffer should increase performance because fewer read syscalls would be made.
+	// If zero is used the system's default buffer length will be used.
+	// ReadBufLen defaults to 32767.
+	ReadBufLen int
+	// Timeout is the length of time to wait before timing out on a read request.
+	// Timeout defaults to nil which means no timeout is used.
+	Timeout *syscall.Timeval
+	// Promisc is set to true for promiscuous mode ethernet sniffing.
+	// Promisc defaults to true.
+	Promisc bool
+	// Immediate is set to true to make our read requests return as soon as a packet becomes available.
+	// Otherwise, a read will block until either the kernel buffer becomes full or a timeout occurs.
+	// The default is true.
+	Immediate bool
+	// PreserveLinkAddr is set to false if the link level source address should be filled in automatically
+	// by the interface output routine. Set to true if the link level source address will be written,
+	// as provided, to the wire.
+	// The default is true.
+	PreserveLinkAddr bool
+}
+
+var defaultOptions = Options{
+	BPFDeviceName:    "",
+	ReadBufLen:       32767,
+	Timeout:          nil,
+	Promisc:          true,
+	Immediate:        true,
+	PreserveLinkAddr: true,
+}
+
+// BPFSniffer is a struct used to track state of a BSD BPF ethernet sniffer
+// such that gopacket's PacketDataSource interface is implemented.
+type BPFSniffer struct {
+	options           *Options
+	sniffDeviceName   string
+	fd                int
+	readBuffer        []byte
+	lastReadLen       int
+	readBytesConsumed int
+}
+
+// NewBPFSniffer is used to create BSD-only BPF ethernet sniffer
+// iface is the network interface device name that you wish to sniff
+// options can set to nil in order to utilize default values for everything.
+// Each field of Options also have a default setting if left unspecified by
+// the user's custome Options struct.
+func NewBPFSniffer(iface string, options *Options) *BPFSniffer {
+	sniffer := BPFSniffer{
+		sniffDeviceName: iface,
+	}
+	if options == nil {
+		sniffer.options = &defaultOptions
+	} else {
+		sniffer.options = options
+	}
+	return &sniffer
+}
+
+// Close is used to close the file-descriptor of the BPF device file.
+func (b *BPFSniffer) Close() error {
+	return syscall.Close(b.fd)
+}
+
+func (b *BPFSniffer) pickBpfDevice() {
+	var err error
+	for i := 0; i < 99; i++ {
+		b.options.BPFDeviceName = fmt.Sprintf("/dev/bpf%d", i)
+		b.fd, err = syscall.Open(b.options.BPFDeviceName, syscall.O_RDWR, 0)
+		if err == nil {
+			break
+		}
+	}
+}
+
+// Init is used to initialize a BPF device for promiscuous sniffing.
+// It also starts a goroutine to continuously read frames.
+func (b *BPFSniffer) Init() error {
```

Comment:
Forgot if I mentioned this earlier, but would it make sense to put this Init logic inside of NewBPFSniffer?

    func NewBPFSniffer(...) (*BPFSniffer, error) {
    }

That way we could avoid people forgetting to call init after New.